Draft PR for the object store repository - only here so that I can add comments.

greschd opened the pull request.

sphuber committed:
Add the base implementation for the new repository

The `Repository` class is the new interface to interact with the file
repository. It contains a file hierarchy in memory to now what is
contained within it. A new instance can be constructed from an existing
hierarchy using the `from_serialized` class method.

The actual storing of the objects in a repository is delegated to a
backend repository instance, represented by `AbstractRepositoryBackend`.
This backend class is only tasked with consuming byte streams and
storing them as objects in some repository under a unique key that
allows the object to be retrieved at a later time. The backend
repository is not expected to keep a mapping of what it contains. It
should merely provide the interface to store byte streams and return
their content, given their unique corresponding key.

This also means that the internal virtual hierarchy of a ``Repository``
instance does not necessarily represent all the files that are stored by
repository backend. The repository exposes a mere subset of all the file
objects stored in the backend. This is why object deletion is also
implemented as a soft delete, by default, where the files are just
removed from the internal virtual hierarchy, but not in the actual
backend. This is because those objects can be referenced by other
instances.

The interface is implemented for the `disk-objectstore` which is an
efficient key-value store on the local file system. This commit also
provides a sandbox implementation which represents a temporary folder on
the local file system.

sphuber committed:
Drop support for Python 3.5

CI testing will now only be done against Python 3.6 and 3.8.

sphuber committed:
Add the `repository_metadata` column to the `Node` database models

This new JSONB column will be used to store the metadata of the contents
of the file repository. It essentially is a mapping of the file object
hierarchy, where each object contains the relative filepath of it within
the repository, which is purely symbolic, and a hashkey, which is the
actual identifier that uniquely identifies the object in the repository
backend.

The column is made nullable and there is no default set. This is to make
sure to not add unnecessary bytes to the database for all the nodes that
contain no files whatsoever, which will be a decent chunk. The front-end
ORM class `Node` will return an empty dictionary if the database model
field is null, such that the clients don't have to deal with `None`.

sphuber committed:
Integrate the new `Repository` implementation

The new `Repository` implementation is integrated by wrapping it in a
new class `NodeRepository` that is mixed into the `Node` class. This
approach serves multiple purposes.

 1. The class serves as a backwards compatibility layer between the old
    repository interface, which would allow clients to set the mode for
    file handles for reading, as well as passing in text-like streams,
    and the new interface, which exclusively operates with bytes. The
    interface will decode the bytes when the caller is expecting normal
    strings in order to not break the interface.

 2. It allows to implement the concept of immutability which is a
    concept of the `Node` and is completely unknown to the `Repository`
    class.

 3. It nicely separates the file repository functionality into a
    separate file preventing the already huge `Node` implementation from
    getting even larger.

The downside of the latter is that the `NodeRepository` class contains
some potentially confusing logic. It accesses `Node` properties, but
from the code it is not clear where they come from. Since this class is
only meant to be used as a mixin for the `Node` class, I think that is
a sacrifice that is worth the benefits.

sphuber committed:
Remove the code to perform incremental backup of the file repository

This code was put in place because a naive backup of the file repository
was impossible already for reasonably size projects. The problem was the
underlying design where each node had an associated folder on disk, even
if it contained no files, and all files were stored separately, creating
a huge numbers of files. This meant that just rsync'ing the folder could
take days even to just calculate the difference with the backed up
folder.

The ad-hoc solution was this script that would only transfer the folders
of the nodes added since the last backup date, the list of which was
determined by a simple query. However, now that the new file repository
is implemented, which has been explicitly designed to be easily backed
up by tools like `rsync`, this custom solution is no longer needed.

sphuber committed:
First implementation of export/import using new disk object store

Use the `Container.export` method in export/import functionality

sphuber committed:
Docs: add "Internal architecture - File repository" section

This section describes in detail the new file repository design,
including the design guidelines that were kept in mind and what
requirements it was designed to respect. It also gives a detailed
overview over how the implementation is integrated in the existing
code base.

sphuber committed:
Add the repository mutating methods to the `Sealable` mixin.

This ensures that the repository is still mutable for sealable nodes as
long as they are not yet sealed. After sealing, the repository of even
these nodes becomes fully immutable.

Note that the implementation completely overrides the methods of the
`NodeRepositoryMixin`, because it needs to override the check in those
that will raise if the node is stored. An alternative would have been to
call the `super` from the `Sealable`, which due to the correct MRO,
would in fact the underlying method on the `NodeRepositoryMixin`, but it
would need to accept an argument, like for example `force`, to skip the
mutability check. This is not desirable, however, since those repository
methods are public methods of the `Node` interface and so any user will
be able to disable the mutability check for any node.

This solution does not suffer from that vulnerability but the downside
is that the implementation of the method from the `NodeRepositoryMixin`
needs to be copied and it needs to be kept in sync.

sphuber committed:
Improve the storage effiency of the `repository_metadata` format

Since now all nodes will have a full representation of the virtual file
hierarchy of their repository content, which is stored in the database
in the `repository_metadata` column, it is crucial that the format of
that metadata is optimized to use as little data as possible, to prevent
the database from bloating unnecessarily.

Currently, we only define two file types:

 * Directories: which can contain an arbitrary number of sub objects
 * Files: which cannot contain objects, but instead have a unique hashkey

This means that we can serialize the metadata, without encoding the
explicit type. If it contains a key, it is necessarily a file. If it is
not a file, it has to be a directory.

The serialization format therefore can make do with simple nested
dictionaries, where each entry either contains the key `k`, indicating
it is a file, or `o`, which can be a dictionary itself, containing the
objects contained within that directory. Note that we can even safely
omit the `o` key for an empty dictionary. Since there is no `k` we know
it has to be a directory anyway.

This logic is implemented in `FileObject.from_serialized` that can fully
reconstruct all instance properties by inferring them from the sparse
serialized data.

sphuber committed:
Make changes backwards compatible

Remaining backward incompatibilities:

 * `Node.open` now is a proper contextmanager as it should be, but that
   means that code calling it straight up with a context, will start to
   raise an exception. Idem dito for `SinglefileData` of course.
 * The `force` argument is removed from various repository modifying
   methods on the `Node` class.
 * The `Repository` class from `aiida.orm.utils.repository` is gone
 * The `_get_base_folder` method on `Repository` and by extension on the
   `Node` class no longer exists, but this is acceptable since it is a
   protected method anyway.

greschd commented on aiida/orm/nodes/data/singlefile.py:
```diff
@@ -56,17 +57,19 @@ def filename(self):
         """Return an open file handle to the content of this data node.
 
-        :param key: optional key within the repository, by default is the `filename` set in the attributes
+        :param path: the relative path of the object within the repository.
```
Renamed `key` -> `path`. In other places a deprecation warning is added for this.

sphuber replied:
Thanks, will add to `UpfData` and `CifData` subclasses as well if necessary

greschd commented on aiida/orm/nodes/data/upf.py:
```diff
@@ -199,10 +200,13 @@ def parse_upf(fname, check_filename=True):
-        fname = fname.name
     except AttributeError:
-        with open(fname, encoding='utf8') as handle:
+        with open(fname) as handle:
```
I think the encoding should still be specified. Note that the default is locale-dependent, so this won't show up in testing - I've been bitten by this more than once :)

sphuber replied:
Is probably the safer choice, except this would still fail if the UPF is not actually utf-8 encoded, and there is no way to specify the correct encoding ....

greschd replied:
Hmm, what about just exposing it in the `parse_upf` interface (and setting the default to `'utf-8'` there)? Or is that not a user-facing function?

sphuber replied:
Could do, not sure if "officially" it is public facing, but people might be using. On the other hand, I have reimplemented it in `aiida-pseudo` so think to just make sure it is done correctly over there.

greschd commented on aiida/backends/general/migrations/utils.py:
```diff
@@ -147,3 +146,100 @@ def recursive_datetime_to_isoformat(value):
     return json.dumps(recursive_datetime_to_isoformat(dictionary))
+
+
+def get_duplicate_uuids(table):
```
These utilities have moved - probably they should be importable through the old location for a transition period. Not really sure if there's anyone outside of `aiida-core` using these, though.

sphuber replied:
These aren't part of the public API, so officially we don't guarantee stability so don't think we need to do anything here

greschd commented on utils/dependency_management.py:
```diff
@@ -30,7 +30,7 @@
 }
 
-CONDA_IGNORE = ['pyblake2', r'.*python_version == \"3\.5\"']
+CONDA_IGNORE = ['pyblake2']
```
Not sure what `CONDA_IGNORE` does, but since `pyblake2` is no longer in the dependencies maybe it's also not needed here?

greschd commented on aiida/cmdline/commands/cmd_node.py:
```diff
@@ -98,18 +98,18 @@ def _copy_tree(key, output_dir):  # pylint: disable=too-many-branches
             # in the AiiDA node, not an actual OS - level path.
             file_key = file.name if not key else key + '/' + file.name
-            if file.type == FileType.DIRECTORY:
+            if file.file_type == FileType.DIRECTORY:
```
Renamed `.type` -> `.file_type`.

sphuber replied:
I have a deprecation for this. The only potential problem is that again (like with the `ExitCode`) I have changed `File` from named tuple into a class. There at least I have backported a `type` property that is deprecated and also constructor accepts `type` forwarding it to `file_type`. Hope that is enough.

greschd replied:
Hmm, yeah I could see someone doing something like 

```python
for name, type in files_list:
    ...
```

Probably more something that would show up in `aiida-core` itself, though.

greschd replied:
Maybe implementing an `__iter__` method (with deprecation warning) would be nice? Something like 

```python
def __iter__(self):
    # print deprecation message
    yield self.name
    yield self.file_type
```

Of course there could be different uses, but that would cover the unpacking use-case.

greschd commented on aiida/orm/nodes/repository.py:
```diff
@@ -0,0 +1,297 @@
+
+        return self._repository.get_object_content(path)
+
+    def put_object_from_filelike(
```
No longer accepts the `force` keyword (probably the same for other methods).

sphuber replied:
Thanks, I have added a deprecation for these.

sphuber commented:
Thanks for going over the changes @greschd . I had most of these but is good to have had a second set of eyes. I will soon open a PR on `aiida-core` to put the deprecations in place. I have another question. Currently the `open` and `get_object_content` methods do allow to specify a `mode`, but they don't allow to specify an `encoding`. I think we should be adding this as well, no? Since this would be adding a keyword, we don't need to do anything now and can just add it to the new interface.

greschd commented:
Yeah, I think we should add an `encoding` option as well. So what happens currently? Is it set to `'utf-8'` somewhere further down, or does it use the system default?

That's probably something to be mindful of, especially for migrating from the old to the new repository. As in, check what happens when migrating if the user has a non-utf8 default encoding.

Note also, obviously `encoding` is not allowed with the binary modes.

sphuber commented:
> Yeah, I think we should add an `encoding` option as well. So what happens currently? Is it set to `'utf-8'` somewhere further down, or does it use the system default?

I think all the way down it will use the default and so fallback on to system's locale default.

> That's probably something to be mindful of, especially for migrating from the old to the new repository. As in, check what happens when migrating if the user has a non-utf8 default encoding.

When we will migrate, we will simply just copy all repository content over in pure bytes. This way we don't have to interpret the bytes and simply not even think about encoding. Works everywhere.

> Note also, obviously `encoding` is not allowed with the binary modes.

Mostly it just doesn't make any sense :) If you request bytes, we're going to give you bytes

greschd commented:
> 
> When we will migrate, we will simply just copy all repository content over in pure bytes. This way we don't have to interpret the bytes and simply not even think about encoding. Works everywhere.
> 

Yeah, but that means we also need to use the system default for the new `open` when encoding is not specified. Otherwise users that _used to_ be able to just `open()` now have to specify their default encoding.
Maybe that's not too bad because it matches what Python does.. but on the other hand, for data sharing it would be _really_ nice to have UTF-8 everywhere.
TBH, maybe it would be a good idea to just let users that have weird encodings bite this particular bullet now.. at some point, this will end up problematic for them anyway.

> 
> Mostly it just doesn't make any sense :) If you request bytes, we're going to give you bytes

We could do 
```python
In [1]: open('foo', mode='wb', encoding='utf-8')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-f5b1bbe0090d> in <module>()
----> 1 open('foo', mode='wb', encoding='utf-8')

ValueError: binary mode doesn't take an encoding argument
```

sphuber commented:
> Maybe that's not too bad because it matches what Python does.. but on the other hand, for data sharing it would be _really_ nice to have UTF-8 everywhere.

We have discussed this problem at great length, but unfortunately, we cannot guarantee this anyway. We can make sure that all files that AiiDA generates (from memory directly) itself are written to disk as utf-8, however, the files that it retrieves for calculation jobs for example, could be in any encoding. Since AiiDA simply retrieves them and stores the bytes, they will have whatever encoding the remote machine wrote them in. This means it is impossible to ensure that all files in the AiiDA repository in the same encoding. Since we can also not deduce it, we decided that the only solution is to provide the user to request bytes or provide the encoding they think files should have, where possible, and just provide sensible defaults. There is not much more we can do

greschd commented:
Aah yeah.. well, it's one of the great problems in computing for a reason 😄

ltalirz committed:
`SlurmScheduler`: always raise for non-zero exit code (#4332)

The `SlurmScheduler` intentionally ignored non-zero exit codes returned
by SLURM when asking the status for a number of job ids. This was put in
place because SLURM will return a non-zero exit code not only in case of
actual errors in attempting to retrieve the status of the requested jobs
but also when specifying just a single job that no longer is active.

Since the latter is not really an error, yet is difficult to distinguish
from a "real" error, the exit code was ignored. However, this could
lead to the plugin sometimes incorrectly ignoring a real problem and
assuming a job was completed when it was in fact still active.

The solution is to use the weird behavior of SLURM that when asking for
more than one job, it will never return a non-zero status, even when
one or more jobs have finished. That is why, when asking for the status
of a single job, we duplicate the job id, such that even when it is no
longer active, the exit status will still be zero.
